### PR TITLE
Remove `sqlx`, support u64 bit integers

### PR DESCRIPTION
**Pull request by adzialocha**
## Description

* Make all sequence numbers and log ids `u64` according to the Bamboo specification
* Re-introduce `BigInt` for the `wasm` interface for `seq_num` and `log_id`, changed APIs accordingly
* Add conversion traits for  `SeqNum`, `LogId` from `&str` and `String` to give a nice API for `aquadoggo` and others to work with string values for large numbers
* Fix issues where large numbers for operations and entries would break in `p2panda-js`, changed APIs accordingly and added tests
* Note: Adding `u64` and `i64` support is easy in the Rust world and CBOR but in JavaScript and JSON it gets complicated, this is why most of the changes here are actually towards supporting large numbers in WASM / JavaScript
* Note: I've made according changes in `beep-boop` and `aquadoggo` to support large integers. See related issues!

**Bonus level**

* Remove `sqlx` from `p2panda-rs`. Since we added `u64` support we put a little bit more work on the backends anyways and it's always been a little ugly to have a notion of databases inside this crate
* Support "float" operation field type in `p2panda-js`
* Add conversion traits from `&str` and `String` to `Hash`

Closes: #174 and #65
Related issues: https://github.com/p2panda/aquadoggo/pull/54 and https://github.com/p2panda/beep-boop/pull/32

## API changes

**Entries**

* Next entry arguments like `seq_num` and `log_id` are received as strings from the JSON RPC endpoint (will be implemented in https://github.com/p2panda/aquadoggo/pull/54). We can't send numbers since they will break for large integers in JSON. Also we can't send `BigInt` instances since they are not supported by JSON. Strings are the only option.
* `signEncodeEntry` takes `BigInt` instances for `seq_num` and `log_id`
* `decodeEntry` returns also `BigInt` instances for them

**Operations**

* Operation fields of type "int" are stored internally as i64 integers but [JS only supports max 53 bits for numbers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER#description). Adding an integer to the operation fields is done with using a string value since I could not find a way to convert from `JsValue(BigInt)` to `i64` easily. This is for the low-level wasm API.
* For the high-level JS API one can add numbers or `BigInt` instances which are then automatically converted into strings.
* When getting an integer from the operation fields one receives a `BigInt` instance for the low-level wasm API, for the high-level JS API one receives a `BigInt` or number when it is below `Number.MAX_SAFE_INTEGER`.

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header
- [x] Check if descriptions and terminology match `handbook` content (and visa-versa)

### CODE REVIEW

File: p2panda-rs/src/wasm/serde.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+use wasm_bindgen::JsValue;
+
+/// Serializes Rust type into JavaScript value (`JsValue`).
+///
+/// Note that this will NOT serialize into a JSON string but an actual JavaScript object
+/// (potentially with non-JSON values inside like "Map" or "BigInt").
+///
+/// This method uses the `serde_wasm_bindgen` serializer, both for bundle size but also for its
+/// support of large number types. With this serializer all i64 and u64 numbers get serialized as
+/// BigInt instances.
```
Comment (sandreae) [inline p2panda-rs/src/wasm/serde.rs:15]: Really useful and well worded description :+1:

File: p2panda-rs/src/wasm/tests/operation.rs [hunk 2]
```diff
@@ -51,48 +49,78 @@ fn operation_fields() {
...
+    );
+
+    // This integer is too large and can't be represented as i64
+    let result = fields.add(
+        "really_big_number".to_string(),
+        "int".to_string(),
+        JsValue::from_str("932187321673219932187732188"),
+    );
     assert!(result.is_err());
 }
```
Comment (sandreae) [inline p2panda-rs/src/wasm/tests/operation.rs:117]: Just so I'm clear, would this work if `JsValue::bigint_from_str("932187321673219932187732188")` was passed as the value instead?

File: p2panda-rs/src/wasm/tests/operation.rs [hunk 3]
```diff
@@ -51,48 +49,78 @@ fn operation_fields() {
...
+    fields
+        .add("age".to_string(), "int".to_string(), JsValue::from_str("5"))
+        .unwrap();
+
+    // "int" fields always get returned as BigInt instances
+    assert_eq!(
+        fields.get("age".to_string()).unwrap(),
+        JsValue::bigint_from_str("5")
+    );
+}
```
Comment (sandreae) [inline p2panda-rs/src/wasm/tests/operation.rs:91]: From reading the PR description I understood int values smaller than `Number.MAX_SAFE_INTEGER` were returned as numbers, not `BigInt`. Looking here that doesn't seem to be right though.

**Review by sandreae — COMMENTED** [review]
This all looks good to me, well done wrapping your head round it all. It's a bit dizzying trying to understand where and how the js number limitations impact the wasm methods.

I have a couple of questios, but it may just be for my own understanding. 

Thanks!

File: p2panda-rs/src/wasm/tests/operation.rs (hunk 2 above)
Comment (adzialocha) [reply p2panda-rs/src/wasm/tests/operation.rs:117]: No, the `add` method only accepts strings for the "int" type. I've tried to use BigInts but I really can't see how to convert it into an i64 except of converting it to a string first and then to an integer which is weird, maybe you have an idea?

File: p2panda-rs/src/wasm/tests/operation.rs (hunk 3 above)
Comment (adzialocha) [reply p2panda-rs/src/wasm/tests/operation.rs:91]: Only for the "high-level" API meaning `p2panda-js`. In WASM (aka "low-level" API) its BigInts :+1:

**Comment by adzialocha** [comment]
I see some bugs in the TypeScript types, will turn this into a draft again, but happy to already receive feedback!

File: p2panda-js/test/fixtures.ts [hunk 4]
```diff
@@ -1,25 +1,51 @@
...
+        ] as OperationValue;
+        fields.set(key, value);
+      });
+
+      operation.fields = fields;
+    }
+
+    return operation;
+  },
+);
```
Comment (adzialocha) [inline p2panda-js/test/fixtures.ts:41]: I have no clue on how to do this in TypeScript (convert from JSON into different type), might need some help here :cry: 

@cafca

**Comment by adzialocha** [comment]
Ok, this should be ready for review now, thanks for the help and feedback so far! This should also be compatible to the regarding changes in `aquadoggo` and `beep-boop` (basically bringing it all to the same level), so please have a look there as well!

File: p2panda-js/src/types.ts [hunk 5]
```diff
@@ -34,11 +34,11 @@ export type EntryRecord = Entry & {
  * Decoded entry containing optional `Operation`.
  */
 export type Entry = {
-  entryHashBacklink: string | null;
-  entryHashSkiplink: string | null;
-  logId: number;
-  operation: Operation | null;
-  seqNum: number;
+  entryHashBacklink: string | undefined;
```
[the remark this answers is not in the document]
Comment (cafca) [reply p2panda-js/src/types.ts:37]: I don't have an answer though

**Review by cafca — APPROVED** [review]

**Pull request merged**